src/ripple/app/misc/impl/AMMHelpers.cpp

```diff
@@ -0,0 +1,206 @@
+{
+    auto const tokens = root2(asset1 * asset2);
+    return toSTAmount(lptIssue, tokens);
+}
+
+/*
+ * Equation 3:
+ * t = T * [(b/B - (sqrt(f2**2 - b/(B*f1)) - f2)) /
+ *          (1 + sqrt(f2**2 - b/(B*f1)) - f2)]
+ * where f1 = 1 - tfee, f2 = (1 - tfee/2)/f1
+ */
+STAmount
+lpTokensIn(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Deposit,
+    STAmount const& lptAMMBalance,
+    std::uint16_t tfee)
+{
+    auto const f1 = feeMult(tfee);
+    auto const f2 = feeMultHalf(tfee) / f1;
+    Number const r = asset1Deposit / asset1Balance;
+    auto const c = root2(f2 * f2 + r / f1) - f2;
+    auto const t = lptAMMBalance * (r - c) / (1 + c);
+    return toSTAmount(lptAMMBalance.issue(), t);
+}
+
+/* Equation 4 solves equation 3 for b:
+ * Let f1 = 1 - tfee, f2 = (1 - tfee/2)/f1, t1 = t/T, t2 = 1 + t1, R = b/B
+ * then
+ * t1 = [R - sqrt(f2**2 + R/f1) + f2] / [1 + sqrt(f2**2 + R/f1] - f2] =>
+ * sqrt(f2**2 + R/f1)*(t1 + 1) = R + f2 + t1*f2 - t1 =>
+ * sqrt(f2**2 + R/f1)*t2 = R + t2*f2 - t1 =>
+ * sqrt(f2**2 + R/f1) = R/t2 + f2 - t1/t2, let d = f2 - t1/t2 =>
+ * sqrt(f2**2 + R/f1) = R/t2 + d =>
+ * f2**2 + R/f1 = (R/t2)**2 +2*d*R/t2 + d**2 =>
+ * (R/t2)**2 + R*(2*d/t2 - 1/f1) + d**2 - f2**2 = 0
+ */
+STAmount
+ammAssetIn(
+    STAmount const& asset1Balance,
+    STAmount const& lptAMMBalance,
+    STAmount const& lpTokens,
+    std::uint16_t tfee)
+{
+    auto const f1 = feeMult(tfee);
+    auto const f2 = feeMultHalf(tfee) / f1;
+    auto const t1 = lpTokens / lptAMMBalance;
+    auto const t2 = 1 + t1;
+    auto const d = f2 - t1 / t2;
+    auto const a = 1 / (t2 * t2);
+    auto const b = 2 * d / t2 - 1 / f1;
+    auto const c = d * d - f2 * f2;
+    return toSTAmount(
+        asset1Balance.issue(), asset1Balance * solveQuadraticEq(a, b, c));
+}
+
+/* Equation 7:
+ * t = T * (c - sqrt(c**2 - 4*R))/2
+ * where R = b/B, c = R*fee + 2 - fee
+ */
+STAmount
+lpTokensOut(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Withdraw,
+    STAmount const& lptAMMBalance,
+    std::uint16_t tfee)
+{
+    Number const fr = asset1Withdraw / asset1Balance;
+    auto const f1 = getFee(tfee);
+    auto const c = fr * f1 + 2 - f1;
+    auto const t = lptAMMBalance * (c - root2(c * c - 4 * fr)) / 2;
+    return toSTAmount(lptAMMBalance.issue(), t);
+}
+
+/* Equation 8 solves equation 7 for b:
+ * c - 2*t/T = sqrt(c**2 - 4*R) =>
+ * c**2 - 4*c*t/T + 4*t**2/T**2 = c**2 - 4*R =>
+ * -4*c*t/T + 4*t**2/T**2 = -4*R =>
+ * -c*t/T + t**2/T**2 = -R -=>
+ * substitute c = R*f + 2 - f =>
+ * -(t/T)*(R*f + 2 - f) + (t/T)**2 = -R, let t1 = t/T =>
+ * -t1*R*f -2*t1 +t1*f +t1**2 = -R =>
+ * R = (t1**2 + t1*(f - 2)) / (t1*f - 1)
+ */
+STAmount
+withdrawByTokens(
+    STAmount const& assetBalance,
+    STAmount const& lptAMMBalance,
+    STAmount const& lpTokens,
+    std::uint16_t tfee)
+{
+    auto const f = getFee(tfee);
+    Number const t1 = lpTokens / lptAMMBalance;
+    auto const b = assetBalance * (t1 * t1 - t1 * (2 - f)) / (t1 * f - 1);
```

Comment:
@gregtatcam, what guarantees that the divisor is not zero (i.e. that `(t1 * f) != 1`) here?

Reply:
`t1*f == 1` => `f*lpTokens/lptAMMBalance == 1` => `f*lpTokens == lptAMMBalance`. `f = fPct/100,000`, where `fPct <= 1,000` (`1%=1,000`) => `f<=0.001`. Since `lpTokens <= lptAMMBalance` => `f*lpTokens != lptAMMBalance`